### PR TITLE
Add support for blkio read/write iops device

### PR DESCRIPTION
**Pull request by Mashimiao**
Signed-off-by: Ma Shimiao mashimiao.fnst@cn.fujitsu.com

### CODE REVIEW

File: daemon/daemon_unix.go [hunk 1]
```diff
@@ -38,6 +38,38 @@ const (
 	platformSupported = true
 )
 
+func getBlkioReadIOpsDevice(config *runconfig.HostConfig) ([]string, error) {
```
Comment (cpuguy83) [inline daemon/daemon_unix.go]: This and `getBlkioWriteIOpsDevice` look exactly the same. Can we instead pass a value instead of runconfig.HostConfig?

Comment (Mashimiao) [inline daemon/daemon_unix.go]: @cpuguy83  fixed.

**Comment by LK4D4** [comment]
@Mashimiao need rebase
LGTM

**Comment by thaJeztah** [comment]
does this depend on the other PR's? https://github.com/docker/docker/issues/14466 and https://github.com/docker/docker/issues/13959?

(looks like this needs a rebase :cry:)

**Comment by LK4D4** [comment]
@Mashimiao Would you mind to answer @thaJeztah question?

**Comment by runcom** [comment]
I think this can go ahead now :)

**Comment by LK4D4** [comment]
@Mashimiao whoa, sorry, need rebase again.

**Comment by Mashimiao** [comment]
This PR is similar to #14460, I want to rebase and fix this after #14460 merged

**Comment by Mashimiao** [comment]
@LK4D4  @runcom  @thaJeztah   PR rebased and fixed. I think it failed with unrelated case.

**Comment by thaJeztah** [comment]
Thanks @Mashimiao, test failure will be fixed in https://github.com/docker/docker/pull/18458

**Comment by thaJeztah** [comment]
https://github.com/docker/docker/pull/18458 was merged, so if you rebase, the tests should be green again

**Comment by Mashimiao** [comment]
Thanks @thaJeztah   rebased.

**Comment by thaJeztah** [comment]
@Mashimiao sorry, looks like this needs a rebase again :cry: could be because of https://github.com/docker/docker/pull/18472 being merged, I hope it isn't bad. Feel free to blame @calavera :smiling_imp:

**Comment by Mashimiao** [comment]
@thaJeztah  That's OK. rebased again.

**Comment by thaJeztah** [comment]
ping @runcom @cpuguy83 PTAL for the latest code changes, let's see if we can get this in docs review now :-)

**Comment by thaJeztah** [comment]
(Thanks @Mashimiao!)

File: daemon/daemon_unix.go [hunk 2]
```diff
@@ -86,6 +86,36 @@ func parseSecurityOpt(container *container.Container, config *runconfig.HostConf
...
+			return nil, err
+		}
+		ReadIOpsDevice := blkiodev.NewThrottleDevice(int64(stat.Rdev/256), int64(stat.Rdev%256), iopsDevice.Rate)
+		BlkioReadIOpsDevice = append(BlkioReadIOpsDevice, ReadIOpsDevice)
+	}
+
+	return BlkioReadIOpsDevice, nil
+}
+
+func getBlkioWriteIOpsDevices(config *runconfig.HostConfig) ([]*blkiodev.ThrottleDevice, error) {
```
Comment (cpuguy83) [inline daemon/daemon_unix.go]: This looks just like `getBlkioReadIOpsDevices` or am I missing something?
Can we just use 1 validation function?

File: daemon/daemon_unix.go [hunk 3]
```diff
@@ -86,6 +86,36 @@ func parseSecurityOpt(container *container.Container, config *runconfig.HostConf
 	return err
 }
 
+func getBlkioReadIOpsDevices(config *runconfig.HostConfig) ([]*blkiodev.ThrottleDevice, error) {
+	var BlkioReadIOpsDevice []*blkiodev.ThrottleDevice
```
Comment (cpuguy83) [inline daemon/daemon_unix.go]: Can we change to `blkioReadIOpsDevice` (lower-cased)

File: daemon/daemon_unix.go (hunk 2 above)
Comment (Mashimiao) [inline daemon/daemon_unix.go]: Yes, they look like same.  But I think it's more clear to use two fucntions for readiops and writeiops.
If you insist, I think using 1 validation function is just OK.

File: daemon/daemon_unix.go (hunk 3 above)
Comment (Mashimiao) [inline daemon/daemon_unix.go]: fixed, thanks.

**Comment by thaJeztah** [comment]
ping @cpuguy83 @LK4D4 @jfrazelle PR was update, PTAL, let's see if we can move this forward \o/

**Comment by jfrazelle** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Comment by LK4D4** [comment]
ping @thaJeztah @moxiegirl

**Comment by calavera** [comment]
Docs LGTM. Unfortunately, this PR will need to be rebased before merging.

@thaJeztah can you take a look?

**Comment by Mashimiao** [comment]
@calavera  thanks, rebased.

File: docs/reference/run.md [hunk 4]
```diff
@@ -999,6 +1001,21 @@ Both flags take limits in the `<device-path>:<limit>[unit]` format. Both read
 and write rates must be a positive integer. You can specify the rate in `kb`
 (kilobytes), `mb` (megabytes), or `gb` (gigabytes).
 
+The `--device-read-iops` flag limits read rate (IO per second) from a device.
+For example, this command creates a container and limits the read rate to
+`1000` per second from `/dev/sda`:
```
Comment (thaJeztah) [inline docs/reference/run.md]: nit: s/`1000` per second/`1000` IO per second/

File: docs/reference/run.md [hunk 5]
```diff
@@ -999,6 +1001,21 @@ Both flags take limits in the `<device-path>:<limit>[unit]` format. Both read
...
+
+    $ docker run -ti --device-read-iops /dev/sda:1000 ubuntu
+
+The `--device-write-iops` flag limits write rate (IO per second) to a device.
+For example, this command creates a container and limits the write rate to
+`1000` per second to `/dev/sda`:
```
Comment (thaJeztah) [inline docs/reference/run.md]: nit: s/`1000` per second/`1000` IO per second/

**Comment by thaJeztah** [comment]
Thanks @Mashimiao two nits. Also can you add the new options to the [API changelog](https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v122-api-changes), for example:

    `POST /containers/create` allows you to set a read/write rate
    limit for a device (in bytes per second or IO per second).

(We forgot to add "BlkioThrottleReadBpsDevice" and "BlkioThrottleWriteBpsDevice" to the change log in https://github.com/docker/docker/pull/14466)

File: docs/reference/run.md (hunk 4 above)
Comment (Mashimiao) [inline docs/reference/run.md]: fixed

File: docs/reference/run.md (hunk 5 above)
Comment (Mashimiao) [inline docs/reference/run.md]: fixed

**Comment by Mashimiao** [comment]
@thaJeztah  added to api Changes

**Comment by Mashimiao** [comment]
jenkins failed with unrelated case

**Comment by thaJeztah** [comment]
Thanks @Mashimiao, happy to see this ready to merge!!

LGTM

Documentation is basically the same as for blkioBps, so it should be good to go. Moving to "merge"

I restarted janky and experimental, but the failing tests are not related, probably a time issue;

```
--- FAIL: TestFormat (0.00s)
	formatter_test.go:152: Expected 
		CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
		containerID1        ubuntu              ""                  45 years ago                                                foobar_baz
		containerID2        ubuntu              ""                  45 years ago                                                foobar_bar
		, got 
		CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
		containerID1        ubuntu              ""                  46 years ago                                                foobar_baz
		containerID2        ubuntu              ""                  46 years ago                                                foobar_bar
FAIL
```

**Comment by runcom** [comment]
seems like failure comes from #18820

**Comment by thaJeztah** [comment]
@runcom yes I just found that one, to check if an issue was already created for it.

**Comment by thaJeztah** [comment]
@Mashimiao can you rebase? Tests should be fixed after that (by https://github.com/docker/docker/pull/18820)

**Comment by thaJeztah** [comment]
It's green! \o/ :tada:

**Pull request merged**